File: sources/Android/android-communications/library/src/test/java/com/polar/androidcommunications/api/ble/model/gatt/client/pmd/model/PmdTimeStampUtilsTest.kt
package com.polar.androidcommunications.api.ble.model.gatt.client.pmd.model

import com.polar.androidcommunications.api.ble.exceptions.NegativeTimeStampError
import com.polar.androidcommunications.api.ble.exceptions.TimeStampAndFrequencyZeroError
import com.polar.androidcommunications.testrules.BleLoggerTestRule
import org.junit.Assert
import org.junit.Assert.assertThrows

import org.junit.Rule
import org.junit.Test
import kotlin.math.round

internal class PmdTimeStampUtilsTest {
    @Rule
    @JvmField
    val bleLoggerTestRule = BleLoggerTestRule()

    @Test
    fun `assert on invalid input, both sample rate and timestamp are zero`() {
        // Arrange
        val previousTimeStamp = 0uL
        val timeStamp = 100000uL //ns
        val samplesSize = 100
        val samplingRate = 0

        // Act &  Assert
        assertThrows(TimeStampAndFrequencyZeroError::class.java) {
            PmdTimeStampUtils.getTimeStamps(
                previousFrameTimeStamp = previousTimeStamp,
                frameTimeStamp = timeStamp,
                samplesSize = samplesSize,
                sampleRate = samplingRate
            )
        }
    }

    @Test
    fun `assert on invalid input, time stamps are negative`() {
        // Arrange
        val previousTimeStamp = 0uL
        val timeStamp = 100000uL //ns
        val samplesSize = 100
        val samplingRate = 52

        // 1/52Hz = 0.019230769230769s = 19230769ns
        // timeStamp - 19230769ns => Exception

        // Act &  Assert
        assertThrows(NegativeTimeStampError::class.java) {
            PmdTimeStampUtils.getTimeStamps(
                previousFrameTimeStamp = previousTimeStamp,
                frameTimeStamp = timeStamp,
                samplesSize = samplesSize,
                sampleRate = samplingRate
            )
        }
    }

    @Test
    fun `samples time stamps based on frequency`() {
        // Arrange
        val previousTimeStamp = 0uL
        val timeStamp = 10000000000uL // 10 seconds
        val samplesSize = 10
        val samplingRate = 1 // 1Hz

        // Act
        val timestamps = PmdTimeStampUtils.getTimeStamps(
            previousFrameTimeStamp = previousTimeStamp,
            frameTimeStamp = timeStamp,
            samplesSize = samplesSize,
            sampleRate = samplingRate
        )
        // Assert
        Assert.assertEquals(samplesSize, timestamps.size)
        Assert.assertEquals(1000000000uL, timestamps[0])
        Assert.assertEquals(2000000000uL, timestamps[1])
        Assert.assertEquals(3000000000uL, timestamps[2])
        Assert.assertEquals(4000000000uL, timestamps[3])
        Assert.assertEquals(5000000000uL, timestamps[4])
        Assert.assertEquals(6000000000uL, timestamps[5])
        Assert.assertEquals(7000000000uL, timestamps[6])
        Assert.assertEquals(8000000000uL, timestamps[7])
        Assert.assertEquals(9000000000uL, timestamps[8])
        Assert.assertEquals(10000000000uL, timestamps[9])
    }

    @Test
    fun `samples time based on frequency only one sample`() {
        // Arrange
        val previousTimeStamp = 0uL
        val timeStamp = 1000000000uL // 1 second
        val samplesSize = 1
        val samplingRate = 1 // 1Hz

        // Act
        val timestamps = PmdTimeStampUtils.getTimeStamps(
            previousFrameTimeStamp = previousTimeStamp,
            frameTimeStamp = timeStamp,
            samplesSize = samplesSize,
            sampleRate = samplingRate
        )
        // Assert
        Assert.assertEquals(samplesSize, timestamps.size)
        Assert.assertEquals(1000000000uL, timestamps[0])
    }

    @Test
    fun `samples time stamps based on previous time stamp test 1`() {
        // Arrange
        val previousTimeStamp = 1000000000uL //1 seconds
        val timeStamp = 11000000000uL // 11 seconds
        val samplesSize = 10
        val samplingRate = 1 // 1Hz

        /*
        Index: prev   0    1    2    3    4    5    6    7    8    9
                 |____|____|____|____|____|____|____|____|____|____|
        Stamp:   1    2    3    4    5    6    7    8    9   10   11
        */
        // Act
        val timestamps = PmdTimeStampUtils.getTimeStamps(
            previousFrameTimeStamp = previousTimeStamp,
            frameTimeStamp = timeStamp,
            samplesSize = samplesSize,
            sampleRate = samplingRate
        )
        // Assert
        Assert.assertEquals(samplesSize, timestamps.size)
        Assert.assertEquals(2000000000uL, timestamps[0])
        Assert.assertEquals(3000000000uL, timestamps[1])
        Assert.assertEquals(4000000000uL, timestamps[2])
        Assert.assertEquals(5000000000uL, timestamps[3])
        Assert.assertEquals(6000000000uL, timestamps[4])
        Assert.assertEquals(7000000000uL, timestamps[5])
        Assert.assertEquals(8000000000uL, timestamps[6])
        Assert.assertEquals(9000000000uL, timestamps[7])
        Assert.assertEquals(10000000000uL, timestamps[8])
        Assert.assertEquals(11000000000uL, timestamps[9])
    }

    @Test
    fun `samples time stamps based on previous time stamp test 2`() {
        // Arrange
        val previousTimeStamp = 100uL // 1 seconds
        val timeStamp = 2000000000uL // 2 seconds
        val samplesSize = 3
        val samplingRate = 1 // 1Hz

        //val startTimeStamp = lastSampleTimeStamp - (timeStampDelta * (samplesSize - 1))

        // (2000000000 - 100) / 3 = 666666633.333333333333333
        /*
        Index: prev                 0               1               2
                 |__666666633.33ns__|__666666633ns__|__666666633ns__|
        Stamp:  100     666666733.33    1333333366.66      2000000000
        */
        // Act
        val timestamps = PmdTimeStampUtils.getTimeStamps(
            previousFrameTimeStamp = previousTimeStamp,
            frameTimeStamp = timeStamp,
            samplesSize = samplesSize,
            sampleRate = samplingRate
        )
        // Assert
        Assert.assertEquals(samplesSize, timestamps.size)
        Assert.assertEquals(round(666666733.33).toULong(), timestamps[0])
        Assert.assertEquals(round(1333333366.66).toULong(), timestamps[1])
        Assert.assertEquals(2000000000uL, timestamps[2])
    }

    @Test
    fun `samples time stamps based on previous time stamp only one sample`() {
        // Arrange
        val previousTimeStamp = 1000000000uL //1 seconds
        val timeStamp = 2000000000uL // 2 seconds
        val samplesSize = 1
        val samplingRate = 1 // 1Hz

        /*
        Index: prev   0
                 |____|
        Stamp:   1    2
        */
        // Act
        val timestamps = PmdTimeStampUtils.getTimeStamps(
            previousFrameTimeStamp = previousTimeStamp,
            frameTimeStamp = timeStamp,
            samplesSize = samplesSize,
            sampleRate = samplingRate
        )
        // Assert
        Assert.assertEquals(samplesSize, timestamps.size)
        Assert.assertEquals(timeStamp, timestamps[0])
    }
}

File: sources/iOS/ios-communications/Tests/iOSCommunicationsTests/PmdTimeStampUtilsTest.swift
//  Copyright © 2022 Polar. All rights reserved.

import XCTest
@testable import iOSCommunications

final class PmdTimeStampUtilsTest: XCTestCase {
       
    func testAssertInvalidInputSampleRateAndTimeStampZero() throws {
        // Arrange
        let previousTimeStamp:UInt64 = 0
        let timeStamp:UInt64 = 100000 //ns
        let samplesSize:UInt = 100
        let samplingRate:UInt = 0
        
        // Act &  Assert
        XCTAssertThrowsError(try PmdTimeStampUtils.getTimeStamps(
            previousFrameTimeStamp: previousTimeStamp,
            frameTimeStamp: timeStamp,
            samplesSize: samplesSize,
            sampleRate: samplingRate
        )) { error in
            guard case BleGattException.gattDataError = error else {
                return XCTFail()
            }
        }
    }
    
    func testAssertInvalidInputTimeStampsGoesNegative() throws {
        // Arrange
        let previousTimeStamp:UInt64 = 0
        let timeStamp:UInt64 = 100000 //ns
        let samplesSize:UInt = 100
        let samplingRate:UInt = 52
        
        // 1/52Hz = 0.019230769230769s = 19230769ns
        // timeStamp - 19230769ns => Exception
        
        // Act & Assert
        XCTAssertThrowsError(try PmdTimeStampUtils.getTimeStamps(
            previousFrameTimeStamp: previousTimeStamp,
            frameTimeStamp: timeStamp,
            samplesSize: samplesSize,
            sampleRate: samplingRate
        )) { error in
            guard case BleGattException.gattDataError = error else {
                return XCTFail()
            }
        }
    }
    
    func testSamplesTimeStampsBasedOnFrequency() throws {
        // Arrange
        let previousTimeStamp:UInt64 = 0
        let timeStamp:UInt64 = 10000000000 // 10 seconds
        let samplesSize:UInt = 10
        let samplingRate:UInt = 1 // 1Hz
        
        // Act
        let timestamps = try PmdTimeStampUtils.getTimeStamps(
            previousFrameTimeStamp: previousTimeStamp,
            frameTimeStamp: timeStamp,
            samplesSize: samplesSize,
            sampleRate: samplingRate
        )
        
        // Assert
        XCTAssertEqual(samplesSize, UInt(timestamps.count))
        XCTAssertEqual(1000000000, timestamps[0])
        XCTAssertEqual(2000000000, timestamps[1])
        XCTAssertEqual(3000000000, timestamps[2])
        XCTAssertEqual(4000000000, timestamps[3])
        XCTAssertEqual(5000000000, timestamps[4])
        XCTAssertEqual(6000000000, timestamps[5])
        XCTAssertEqual(7000000000, timestamps[6])
        XCTAssertEqual(8000000000, timestamps[7])
        XCTAssertEqual(9000000000, timestamps[8])
        XCTAssertEqual(10000000000, timestamps[9])
    }
    
    func testSamplesTimeBasedOnFrequencyOnlyOneSample() throws {
        // Arrange
        let previousTimeStamp:UInt64 = 0
        let timeStamp:UInt64 = 1000000000 // 1 second
        let samplesSize:UInt = 1
        let samplingRate:UInt = 1 // 1Hz
        
        // Act
        let timestamps = try PmdTimeStampUtils.getTimeStamps(
            previousFrameTimeStamp: previousTimeStamp,
            frameTimeStamp: timeStamp,
            samplesSize: samplesSize,
            sampleRate: samplingRate
        )
        
        // Assert
        XCTAssertEqual(samplesSize, UInt(timestamps.count))
        XCTAssertEqual(1000000000, timestamps[0])
    }
    
    func testSamplesTimeStampsBasedOnPreviousTimeStampTest1() throws {
        // Arrange
        let previousTimeStamp:UInt64 = 1000000000 // 1 second
        let timeStamp:UInt64 = 11000000000 // 11 seconds
        let samplesSize:UInt = 10
        let samplingRate:UInt = 1 // 1Hz
        
        /*
         Index: prev   0    1    2    3    4    5    6    7    8    9
         |____|____|____|____|____|____|____|____|____|____|
         Stamp:   1    2    3    4    5    6    7    8    9   10   11
         */
        // Act
        let timestamps = try PmdTimeStampUtils.getTimeStamps(
            previousFrameTimeStamp: previousTimeStamp,
            frameTimeStamp: timeStamp,
            samplesSize: samplesSize,
            sampleRate: samplingRate
        )
        
        // Assert
        XCTAssertEqual(samplesSize,UInt(timestamps.count))
        XCTAssertEqual(2000000000, timestamps[0])
        XCTAssertEqual(3000000000, timestamps[1])
        XCTAssertEqual(4000000000, timestamps[2])
        XCTAssertEqual(5000000000, timestamps[3])
        XCTAssertEqual(6000000000, timestamps[4])
        XCTAssertEqual(7000000000, timestamps[5])
        XCTAssertEqual(8000000000, timestamps[6])
        XCTAssertEqual(9000000000, timestamps[7])
        XCTAssertEqual(10000000000, timestamps[8])
        XCTAssertEqual(11000000000, timestamps[9])
    }
    
    func testSamplesTimeStampsBasedOnPreviousTimeStampTest2() throws {
        // Arrange
        let previousTimeStamp:UInt64 = 100
        let timeStamp:UInt64 = 2000000000 // 2 seconds
        let samplesSize:UInt = 3
        let samplingRate:UInt = 1 // 1Hz
                
        //val startTimeStamp = lastSampleTimeStamp - (timeStampDelta * (samplesSize - 1))
        
        // (2000000000 - 100) / 3 = 666666633.333333333333333
        /*
         Index: prev                 0               1               2
         |__666666633.33ns__|__666666633ns__|__666666633ns__|
         Stamp:  100     666666733.33    1333333366.66      2000000000
         */
        
        // Act
        let timestamps = try PmdTimeStampUtils.getTimeStamps(
            previousFrameTimeStamp: previousTimeStamp,
            frameTimeStamp: timeStamp,
            samplesSize: samplesSize,
            sampleRate: samplingRate
        )
        
        // Assert
        XCTAssertEqual(samplesSize, UInt(timestamps.count))
        XCTAssertEqual(UInt64(round(666666733.33)), timestamps[0])
        XCTAssertEqual(UInt64(round(1333333366.66)), timestamps[1])
        XCTAssertEqual(2000000000, timestamps[2])
    }
    
    func testSamplesTimeStampsBasedOnPreviousTimeStampOnlyOneSample() throws {
        // Arrange
        let previousTimeStamp:UInt64 = 1000000000 //1 second
        let timeStamp:UInt64 = 2000000000 // 2 seconds
        let samplesSize:UInt = 1
        let samplingRate:UInt = 1 // 1Hz
        
        
        /*
         Index: prev   0
         |____|
         Stamp:   1    2
         */
        
        // Act
        let timestamps = try PmdTimeStampUtils.getTimeStamps(
            previousFrameTimeStamp: previousTimeStamp,
            frameTimeStamp: timeStamp,
            samplesSize: samplesSize,
            sampleRate: samplingRate
        )
        // Assert
        XCTAssertEqual(samplesSize, UInt(timestamps.count))
        XCTAssertEqual(timeStamp, timestamps[0])
    }
    
    func testPerformanceExample() throws {
        // Arrange
        let previousTimeStamp:UInt64 = 0
        let timeStamp:UInt64 = 10000000000 // 10 seconds
        let samplesSize:UInt = 100
        let samplingRate:UInt = 100 // 1Hz
        
        
        self.measure {
            // Act
            do {
                try PmdTimeStampUtils.getTimeStamps(
                    previousFrameTimeStamp: previousTimeStamp,
                    frameTimeStamp: timeStamp,
                    samplesSize: samplesSize,
                    sampleRate: samplingRate
                )
            } catch {
                print("catch")
            }
        }
    }
}
